src/PackagePrivate/DatabaseTermIdsResolver.php

```diff
@@ -0,0 +1,129 @@
+
+	/** @var ILoadBalancer */
+	private $lb;
+
+	/** @var IDatabase */
+	private $dbr = null;
+
+	/** @var IDatabase */
+	private $dbw = null;
+
+	/** @var string[] stash of data returned from the {@link TypeIdsResolver} */
+	private $typeNames = [];
+
+	public function __construct(
+		TypeIdsResolver $typeIdsResolver,
+		ILoadBalancer $lb
+	) {
+		$this->typeIdsResolver = $typeIdsResolver;
+		$this->lb = $lb;
+	}
+
+	/*
+	 * Term data is first read from the replica; if that returns less rows than we asked for,
+	 * then there are some new rows in the master that were not yet replicated, and we fall back
+	 * to the master. As the internal relations of the term store never change (for example,
+	 * a term_in_lang row will never suddenly point to a different text_in_lang), a master fallback
+	 * should never be necessary in any other case. However, callers need to consider where they
+	 * got the list of term IDs they pass into this method from: if it’s from a replica, they may
+	 * still see outdated data overall.
+	 */
+	public function resolveTermIds( array $termIds ): array {
+		$terms = [];
+		$this->connectDbr();
+
+		$replicaResult = $this->selectTerms( $this->dbr, $termIds );
+		$this->preloadTypes( $replicaResult );
+		$replicaTermIds = [];
+
+		foreach ( $replicaResult as $row ) {
+			$replicaTermIds[] = $row->wbtl_id;
+			$this->addResultTerms( $terms, $row );
+		}
+
+		if ( count( $replicaTermIds ) !== count( $termIds ) ) {
+			$masterTermIds = array_values( array_diff( $termIds, $replicaTermIds ) );
+			$this->connectDbw();
+			$masterResult = $this->selectTerms( $this->dbw, $masterTermIds );
+			$this->preloadTypes( $masterResult );
+			foreach ( $masterResult as $row ) {
+				$this->addResultTerms( $terms, $row );
+			}
+		}
+
+		return $terms;
+	}
+
+	private function selectTerms( IDatabase $db, array $termIds ): IResultWrapper {
+		return $db->select(
+			[ 'wbt_term_in_lang', 'wbt_text_in_lang', 'wbt_text' ],
+			[ 'wbtl_id', 'wbtl_type_id', 'wbxl_language', 'wbx_text' ],
+			[
+				'wbtl_id' => $termIds,
+				// join conditions
+				'wbtl_text_in_lang_id=wbxl_id',
+				'wbxl_text_id=wbx_id',
+			],
+			__METHOD__
+		);
+	}
+
+	private function preloadTypes( IResultWrapper $result ) {
+		$typeIds = [];
+		foreach ( $result as $row ) {
+			$typeId = $row->wbtl_type_id;
+			if ( !array_key_exists( $typeId, $this->typeNames ) ) {
+				$typeIds[$typeId] = true;
+			}
+		}
+		$this->typeNames += $this->typeIdsResolver->resolveTypeIds( array_keys( $typeIds ) );
+	}
+
+	private function addResultTerms( array &$terms, stdClass $row ) {
+		$type = $this->lookupType( $row->wbtl_type_id );
+		$lang = $row->wbxl_language;
+		$text = $row->wbx_text;
+		$terms[$type][$lang][] = $text;
+	}
+
+	private function lookupType( $typeId ) {
+		$typeName = $this->typeNames[$typeId] ?? null;
+		if ( $typeName === null ) {
+			throw new InvalidArgumentException(
+				'Type ID ' . $typeId . ' was requested but not preloaded!' );
```

Comment:
suggested change is included in #34 (except for selecting `wbtl_id` and including it in the message. that was left out from that PR), since I had to touch this part there anyway